diff --git a/Cargo.toml b/Cargo.toml
index 60391ab..ce1532f 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -26,7 +26,7 @@ rustix = { version = "1.0.1", features = ["fs", "event", "net", "time", "mm"] }
 linux-raw-sys = { version = "0.9.2", features = ["ioctl"] }
 
 [target.'cfg(target_os="windows")'.dependencies]
-windows-sys = { version = "0.59.0", features = ["Win32_Devices_Usb", "Win32_Devices_DeviceAndDriverInstallation", "Win32_Foundation", "Win32_Devices_Properties", "Win32_Storage_FileSystem", "Win32_Security", "Win32_System_IO", "Win32_System_Registry", "Win32_System_Com"] }
+windows-sys = { version = "0.59.0", features = ["Win32_Devices_Usb", "Win32_Devices_DeviceAndDriverInstallation", "Win32_Foundation", "Win32_Devices_Properties", "Win32_Storage_FileSystem", "Win32_Security", "Win32_System_IO", "Win32_System_Ioctl", "Win32_System_Registry", "Win32_System_Com"] }
 
 [target.'cfg(target_os="macos")'.dependencies]
 core-foundation = "0.9.3"
diff --git a/src/hotplug.rs b/src/hotplug.rs
index 1b5ce1a..7376437 100644
--- a/src/hotplug.rs
+++ b/src/hotplug.rs
@@ -31,6 +31,9 @@ pub enum HotplugEvent {
     /// A device has been connected.
     Connected(DeviceInfo),
 
+    /// Serial device connected
+    SerialConnected(DeviceInfo),
+
     /// A device has been disconnected.
     Disconnected(DeviceId),
 }
diff --git a/src/platform/macos_iokit/hotplug.rs b/src/platform/macos_iokit/hotplug.rs
index f95444b..aee0cb4 100644
--- a/src/platform/macos_iokit/hotplug.rs
+++ b/src/platform/macos_iokit/hotplug.rs
@@ -6,6 +6,7 @@ use std::{
 };
 
 use core_foundation::{base::TCFType, dictionary::CFDictionary, runloop::CFRunLoopSource};
+use io_kit_sys::serial::keys::kIOSerialBSDServiceValue;
 use io_kit_sys::{
     kIOMasterPortDefault,
     keys::{kIOFirstMatchNotification, kIOTerminatedNotification},
@@ -52,6 +53,7 @@ pub(crate) struct MacHotplugWatch {
     waker_id: SlabWaker,
     terminated_iter: IoServiceIterator,
     matched_iter: IoServiceIterator,
+    cdc_matched_iter: IoServiceIterator,
     _registration: EventRegistration,
     _notification_port: NotificationPort,
 }
@@ -84,6 +86,14 @@ impl MacHotplugWatch {
             CFDictionary::wrap_under_create_rule(d)
         };
 
+        let cdc_dictionary = unsafe {
+            let d = IOServiceMatching(kIOSerialBSDServiceValue);
+            if d.is_null() {
+                return Err(Error::other("IOServiceMatching failed"));
+            }
+            CFDictionary::wrap_under_create_rule(d)
+        };
+
         let notification_port = NotificationPort::new();
         let terminated_iter = register_notification(
             &notification_port,
@@ -97,6 +107,12 @@ impl MacHotplugWatch {
             &waker_id,
             kIOFirstMatchNotification,
         )?;
+        let cdc_matched_iter = register_notification(
+            &notification_port,
+            &cdc_dictionary,
+            &waker_id,
+            kIOFirstMatchNotification,
+        )?;
 
         let source = unsafe {
             CFRunLoopSource::wrap_under_get_rule(IONotificationPortGetRunLoopSource(
@@ -109,6 +125,7 @@ impl MacHotplugWatch {
             waker_id,
             terminated_iter,
             matched_iter,
+            cdc_matched_iter,
             _registration: registration,
             _notification_port: notification_port,
         })
@@ -125,6 +142,14 @@ impl MacHotplugWatch {
             }
         }
 
+        while let Some(s) = self.cdc_matched_iter.next() {
+            if let Some(dev) = probe_device(s) {
+                return Poll::Ready(HotplugEvent::SerialConnected(dev));
+            } else {
+                debug!("failed to probe serial connected device");
+            }
+        }
+
         if let Some(s) = self.terminated_iter.next() {
             if let Some(registry_id) = get_registry_id(&s) {
                 debug!("device {registry_id} disconnected");
diff --git a/src/platform/windows_winusb/hotplug.rs b/src/platform/windows_winusb/hotplug.rs
index 491d0ad..b1ffb73 100644
--- a/src/platform/windows_winusb/hotplug.rs
+++ b/src/platform/windows_winusb/hotplug.rs
@@ -20,6 +20,7 @@ use windows_sys::Win32::{
         Usb::GUID_DEVINTERFACE_USB_DEVICE,
     },
     Foundation::ERROR_SUCCESS,
+    System::Ioctl::GUID_DEVINTERFACE_COMPORT,
 };
 
 use crate::{
@@ -35,6 +36,7 @@ use super::DevInst;
 pub(crate) struct WindowsHotplugWatch {
     inner: *mut HotplugInner,
     registration: HCMNOTIFICATION,
+    registration_com: HCMNOTIFICATION,
 }
 
 struct HotplugInner {
@@ -45,6 +47,7 @@ struct HotplugInner {
 #[derive(Debug)]
 enum Action {
     Connect,
+    SerialConnect,
     Disconnect,
 }
 
@@ -82,9 +85,37 @@ impl WindowsHotplugWatch {
             return Err(Error::other("Failed to initialize hotplug notifications"));
         }
 
+        let mut registration_com = ptr::null_mut();
+        let filter_com = CM_NOTIFY_FILTER {
+            cbSize: size_of::<CM_NOTIFY_FILTER>() as u32,
+            Flags: 0,
+            FilterType: CM_NOTIFY_FILTER_TYPE_DEVICEINTERFACE,
+            Reserved: 0,
+            u: CM_NOTIFY_FILTER_0 {
+                DeviceInterface: CM_NOTIFY_FILTER_0_2 {
+                    ClassGuid: GUID_DEVINTERFACE_COMPORT,
+                },
+            },
+        };
+
+        let cr_com = unsafe {
+            CM_Register_Notification(
+                &filter_com,
+                inner as *mut c_void,
+                Some(hotplug_callback_serial),
+                &mut registration_com,
+            )
+        };
+
+        if cr_com != CR_SUCCESS {
+            error!("CM_Register_Notification for comport failed: {cr}");
+            return Err(Error::other("Failed to initialize hotplug notifications"));
+        }
+        
         Ok(WindowsHotplugWatch {
             inner,
             registration,
+            registration_com,
         })
     }
 
@@ -105,6 +136,11 @@ impl WindowsHotplugWatch {
                     return Poll::Ready(HotplugEvent::Connected(dev));
                 };
             }
+            Some((Action::SerialConnect, devinst)) => {
+                if let Some(dev) = probe_device(devinst) {
+                    return Poll::Ready(HotplugEvent::SerialConnected(dev));
+                };
+            }
             Some((Action::Disconnect, devinst)) => {
                 return Poll::Ready(HotplugEvent::Disconnected(DeviceId(devinst)));
             }
@@ -128,6 +164,7 @@ impl Drop for WindowsHotplugWatch {
             // immediately afterward without races.
             // [1]: https://learn.microsoft.com/en-us/windows/win32/api/cfgmgr32/nf-cfgmgr32-cm_unregister_notification
             CM_Unregister_Notification(self.registration);
+            CM_Unregister_Notification(self.registration_com);
             drop(Box::from_raw(self.inner));
         }
     }
@@ -167,3 +204,38 @@ unsafe extern "system" fn hotplug_callback(
 
     ERROR_SUCCESS
 }
+
+unsafe extern "system" fn hotplug_callback_serial(
+    _hnotify: HCMNOTIFICATION,
+    context: *const ::core::ffi::c_void,
+    action: CM_NOTIFY_ACTION,
+    eventdata: *const CM_NOTIFY_EVENT_DATA,
+    _eventdatasize: u32,
+) -> u32 {
+    let inner = unsafe { &*(context as *const HotplugInner) };
+
+    let action = match action {
+        CM_NOTIFY_ACTION_DEVICEINTERFACEARRIVAL => Action::SerialConnect,
+        _ => {
+            debug!("Hotplug callback: unknown or unhandled action {action}");
+            return ERROR_SUCCESS;
+        }
+    };
+
+    let device_interface =
+        unsafe { WCStr::from_ptr(addr_of!((*eventdata).u.DeviceInterface.SymbolicLink[0])) };
+
+    let device_instance =
+        get_device_interface_property::<WCString>(device_interface, DEVPKEY_Device_InstanceId)
+            .unwrap();
+    let devinst = DevInst::from_instance_id(&device_instance).unwrap();
+
+    if let Some(parent) = devinst.parent() {
+        debug!("Hotplug com callback: action={action:?}, instance={device_instance}");
+        inner.events.lock().unwrap().push_back((action, parent));
+        inner.waker.wake();
+    } else {
+        debug!("Hotplug com callback no parent devinst found: action={action:?}, instance={device_instance}");
+    }
+    return ERROR_SUCCESS;
+}
